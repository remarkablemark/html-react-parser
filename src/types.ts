import type { DomHandlerOptions } from 'domhandler';
import type { DOMNode } from 'html-dom-parser';
import type { ParserOptions } from 'htmlparser2';
import type { JSX, ReactNode } from 'react';

export interface HTMLReactParserOptions {
  htmlparser2?: ParserOptions & DomHandlerOptions;

  library?: {
    /* eslint-disable @typescript-eslint/no-explicit-any */
    cloneElement: (
      element: JSX.Element,
      props?: object,
      ...children: any[]
    ) => JSX.Element;

    createElement: (
      type: any,
      props?: object,
      ...children: any[]
    ) => JSX.Element;

    isValidElement: (element: any) => boolean;

    [key: string]: any;
    /* eslint-enable @typescript-eslint/no-explicit-any */
  };

  replace?: (
    domNode: DOMNode,
    index: number,
  ) => JSX.Element | string | null | boolean | object | void;

  transform?: (
    reactNode: ReactNode,
    domNode: DOMNode,
    index: number,
  ) => JSX.Element | string | null | void;

  trim?: boolean;
}
