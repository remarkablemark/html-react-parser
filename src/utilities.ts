import { version } from 'react';
import StyleToJS from 'style-to-js';
import type { Element } from 'html-dom-parser';

import type { Props } from './attributes-to-props';

const RESERVED_SVG_MATHML_ELEMENTS = new Set([
  'annotation-xml',
  'color-profile',
  'font-face',
  'font-face-src',
  'font-face-uri',
  'font-face-format',
  'font-face-name',
  'missing-glyph',
] as const);

type ReservedSvgMathmlElements =
  typeof RESERVED_SVG_MATHML_ELEMENTS extends Set<infer T> ? T : never;

/**
 * Check if a tag is a custom component.
 *
 * @see {@link https://github.com/facebook/react/blob/v16.6.3/packages/react-dom/src/shared/isCustomComponent.js}
 *
 * @param tagName - Tag name.
 * @param props - Props passed to the element.
 * @returns - Whether the tag is custom component.
 */
export function isCustomComponent(
  tagName: string,
  props?: Record<PropertyKey, string>,
): boolean {
  if (!tagName.includes('-')) {
    return Boolean(props && typeof props.is === 'string');
  }

  // These are reserved SVG and MathML elements.
  // We don't mind this whitelist too much because we expect it to never grow.
  // The alternative is to track the namespace in a few places which is convoluted.
  // https://w3c.github.io/webcomponents/spec/custom/#custom-elements-core-concepts
  if (RESERVED_SVG_MATHML_ELEMENTS.has(tagName as ReservedSvgMathmlElements)) {
    return false;
  }

  return true;
}

const styleOptions = {
  reactCompat: true,
} as const;

/**
 * Sets style prop.
 *
 * @param style - Inline style.
 * @param props - Props object.
 */
export function setStyleProp(style: string, props: Props): void {
  if (typeof style !== 'string') {
    return;
  }

  if (!style.trim()) {
    props.style = {};
    return;
  }

  try {
    props.style = StyleToJS(style, styleOptions);
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
  } catch (error) {
    props.style = {};
  }
}

/**
 * @see https://reactjs.org/blog/2017/09/08/dom-attributes-in-react-16.html
 */
export const PRESERVE_CUSTOM_ATTRIBUTES = Number(version.split('.')[0]) >= 16;

/**
 * @see https://github.com/facebook/react/blob/cae635054e17a6f107a39d328649137b83f25972/packages/react-dom/src/client/validateDOMNesting.js#L213
 */
export const ELEMENTS_WITH_NO_TEXT_CHILDREN = new Set([
  'tr',
  'tbody',
  'thead',
  'tfoot',
  'colgroup',
  'table',
  'head',
  'html',
  'frameset',
] as const);

type ElementsWithNoTextChildren =
  typeof ELEMENTS_WITH_NO_TEXT_CHILDREN extends Set<infer T> ? T : never;

/**
 * Checks if the given node can contain text nodes
 *
 * @param node - Element node.
 * @returns - Whether the node can contain text nodes.
 */
export const canTextBeChildOfNode = (node: Element) =>
  !ELEMENTS_WITH_NO_TEXT_CHILDREN.has(node.name as ElementsWithNoTextChildren);

/**
 * Returns the first argument as is.
 *
 * @param arg - The argument to be returned.
 * @returns - The input argument `arg`.
 */
export const returnFirstArg = (arg: any) => arg;
