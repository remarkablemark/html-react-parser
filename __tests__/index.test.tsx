import * as domhandler from 'domhandler';
import type { Element } from 'html-dom-parser';
import type { JSX } from 'react';

import * as HTMLReactParser from '../src';
import parse from '../src';
import { html, svg } from './data';
import { render } from './helpers';

describe('module', () => {
  it('exports default', () => {
    expect(parse).toBeInstanceOf(Function);
  });

  it.each(['default', 'attributesToProps', 'domToReact', 'htmlToDOM'] as const)(
    'exports %p',
    (key) => {
      expect(HTMLReactParser[key]).toBeInstanceOf(Function);
    },
  );

  it.each(['Comment', 'Element', 'ProcessingInstruction', 'Text'] as const)(
    'exports %s',
    (key) => {
      expect(HTMLReactParser[key]).toBeInstanceOf(Function);
      expect(HTMLReactParser[key]).toBe(
        domhandler[key as keyof typeof domhandler],
      );
    },
  );
});

describe('HTMLReactParser', () => {
  it.each([undefined, null, {}, [], true, false, 0, 1, () => {}, new Date()])(
    'throws error for value: %p',
    (value) => {
      expect(() => {
        parse(value as string);
      }).toThrow(TypeError);
    },
  );

  it('parses empty string to empty array', () => {
    expect(parse('')).toEqual([]);
  });

  it.each(['a', 'text'])('parses string', (text) => {
    expect(parse(text)).toBe(text);
  });

  it.each(['\n', '\r', '\n\r', 'foo\nbar', 'foo\rbar', 'foo\n\rbar\r'])(
    'parses string with newlines %p',
    (text) => {
      expect(parse(text)).toBe(text);
    },
  );

  it.each([
    '\n<br>',
    '<br>\r',
    '\n<br>\r',
    '<p>foo\nbar\r</p>',
    '<p>foo</p>\rbar',
    'foo<p>\n\rbar</p>\r',
  ])('parses HTML with newlines', (html) => {
    expect(parse(html)).toMatchSnapshot();
  });

  it('parses single HTML element', () => {
    expect(parse(html.single)).toMatchSnapshot();
  });

  it('parses single HTML element with comment', () => {
    // comment should be ignored
    expect(parse(html.single + html.comment)).toMatchSnapshot();
  });

  it('parses multiple HTML elements', () => {
    expect(parse(html.multiple)).toMatchSnapshot();
  });

  it('parses complex HTML with doctype', () => {
    expect(parse(html.doctype + html.complex)).toMatchSnapshot();
  });

  it('parses empty <script>', () => {
    expect(parse('<script></script>')).toMatchSnapshot();
  });

  it('parses empty <style>', () => {
    expect(parse('<style></style>')).toMatchSnapshot();
  });

  it('parses form', () => {
    expect(parse(html.form)).toMatchSnapshot();
  });

  it('parses list', () => {
    expect(parse(html.list)).toMatchSnapshot();
  });

  it('parses template', () => {
    expect(parse(html.template)).toMatchSnapshot();
  });

  it('parses SVG', () => {
    expect(parse(svg.complex)).toMatchSnapshot();
  });

  it('decodes HTML entities', () => {
    const encodedEntities = 'asdf &amp; &yuml; &uuml; &apos;';
    const decodedEntities = "asdf & ÿ ü '";
    const reactElement = parse('<i>' + encodedEntities + '</i>') as JSX.Element;
    expect(reactElement.props.children).toBe(decodedEntities);
  });

  it('escapes tags inside of <title>', () => {
    expect(parse(html.title)).toMatchSnapshot();
  });
});

describe('replace option', () => {
  it('replaces the element if a valid React element is returned', () => {
    expect(
      parse(html.complex, {
        replace(domNode) {
          if ((domNode as Element).name === 'title') {
            return <title>Replaced Title</title>;
          }
        },
      }),
    ).toMatchSnapshot();
  });

  it('does not replace the element if an invalid React element is returned', () => {
    expect(
      parse(html.complex, {
        replace(domNode) {
          if ((domNode as Element).attribs?.id === 'header') {
            return {
              type: 'h1',
              props: { children: 'Heading' },
            };
          }
        },
      }),
    ).toMatchSnapshot();
  });
});

describe('library option', () => {
  const Preact = require('preact');
  const React = require('react');

  it('converts with Preact instead of React', () => {
    const parsedElement = parse(html.single, { library: Preact });
    const preactElement = Preact.createElement('p', {}, 'foo');
    expect(React.isValidElement(parsedElement)).toBe(false);
    expect(Preact.isValidElement(parsedElement)).toBe(true);
    // remove `__v` key since it's causing test equality to fail
    // @ts-expect-error Property '__v' does not exist on type 'string | Element | Element[]'. Property '__v' does not exist on type 'string'.
    delete parsedElement.__v;
    delete preactElement.__v;
    expect(parsedElement).toEqual(preactElement);
  });
});

describe('htmlparser2 option', () => {
  it('parses XHTML with xmlMode enabled', () => {
    // using self-closing syntax (`/>`) for non-void elements is invalid
    // which causes elements to nest instead of being rendered correctly
    // enabling htmlparser2 option xmlMode resolves this issue
    const options = { htmlparser2: { xmlMode: true } };
    expect(parse('<ul><li/><li/></ul>', options)).toMatchSnapshot();
  });
});

describe('trim option', () => {
  it('preserves whitespace text nodes when disabled if valid in parent (default)', () => {
    const html = `<table>
  <tbody>
    <tr><td>hello</td><td>\n</td><td>&nbsp;</td>\t</tr>\r
  </tbody>
</table>`;
    const reactElement = parse(html) as JSX.Element;
    expect(render(reactElement)).toBe(
      '<table><tbody><tr><td>hello</td><td>\n</td><td>\u00a0</td></tr></tbody></table>',
    );
    expect(reactElement).toMatchSnapshot();
  });

  it('removes whitespace text nodes when enabled', () => {
    const html = `<table>
      <tbody><tr><td> text </td><td> </td>\t</tr>\r</tbody>\n</table>`;
    const options = { trim: true };
    const reactElement = parse(html, options) as JSX.Element;
    expect(render(reactElement)).toBe(
      '<table><tbody><tr><td> text </td><td></td></tr></tbody></table>',
    );
  });
});

describe('invalid styles', () => {
  it('copes with invalid styles', () => {
    const html = '<p style="font - size: 1em">X</p>';
    const reactElement = parse(html) as JSX.Element;
    expect(render(reactElement)).toBe('<p>X</p>');
  });
});
