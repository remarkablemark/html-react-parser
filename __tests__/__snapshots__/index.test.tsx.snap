// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTMLReactParser escapes tags inside of <title> 1`] = `
<title>
  &lt;em&gt;text&lt;/em&gt;
</title>
`;

exports[`HTMLReactParser parses HTML with newlines 1`] = `
[
  "
",
  <br />,
]
`;

exports[`HTMLReactParser parses HTML with newlines 2`] = `
[
  <br />,
  "
",
]
`;

exports[`HTMLReactParser parses HTML with newlines 3`] = `
[
  "
",
  <br />,
  "
",
]
`;

exports[`HTMLReactParser parses HTML with newlines 4`] = `
<p>
  foo
bar
</p>
`;

exports[`HTMLReactParser parses HTML with newlines 5`] = `
[
  <p>
    foo
  </p>,
  "
bar",
]
`;

exports[`HTMLReactParser parses HTML with newlines 6`] = `
[
  "foo",
  <p>
    

bar
  </p>,
  "
",
]
`;

exports[`HTMLReactParser parses SVG 1`] = `
<svg
  height="400"
  width="450"
>
  <path
    d="M 100 350 l 150 -300"
    fill="none"
    id="lineAB"
    stroke="red"
    strokeWidth="3"
  />
  <g
    fill="black"
    stroke="black"
    strokeWidth="3"
  >
    <circle
      cx="100"
      cy="350"
      id="pointA"
      r="3"
    />
  </g>
  <g
    fill="black"
    fontFamily="sans-serif"
    fontSize="30"
    stroke="none"
    textAnchor="middle"
  >
    <text
      dx="-30"
      x="100"
      y="350"
    >
      A
    </text>
  </g>
  Your browser does not support inline SVG.
</svg>
`;

exports[`HTMLReactParser parses complex HTML with doctype 1`] = `
<html>
  <head>
    <meta
      charSet="utf-8"
    />
    <title>
      Title
    </title>
    <link
      href="style.css"
      rel="stylesheet"
    />
  </head>
  <body>
    <header
      id="header"
    >
      Header
    </header>
    <h1
      style={
        {
          "color": "#000",
          "fontSize": "42px",
        }
      }
    >
      Heading
    </h1>
    <hr />
    <p>
      Paragraph
    </p>
    <img
      src="image.jpg"
    />
    <div
      className="class1 class2"
    >
      Some 
      <em>
        text
      </em>
      .
    </div>
    <script
      dangerouslySetInnerHTML={
        {
          "__html": "alert();",
        }
      }
    />
  </body>
</html>
`;

exports[`HTMLReactParser parses empty <script> 1`] = `<script />`;

exports[`HTMLReactParser parses empty <style> 1`] = `<style />`;

exports[`HTMLReactParser parses form 1`] = `
<input
  defaultChecked={true}
  defaultValue="foo"
  type="text"
/>
`;

exports[`HTMLReactParser parses list 1`] = `
<ol>
  <li>
    One
  </li>
  <li
    value="2"
  >
    Two
  </li>
</ol>
`;

exports[`HTMLReactParser parses multiple HTML elements 1`] = `
[
  <p>
    foo
  </p>,
  <p>
    bar
  </p>,
]
`;

exports[`HTMLReactParser parses single HTML element 1`] = `
<p>
  foo
</p>
`;

exports[`HTMLReactParser parses single HTML element with comment 1`] = `
<p>
  foo
</p>
`;

exports[`HTMLReactParser parses template 1`] = `
<template>
  <article>
    <p>
      Test
    </p>
  </article>
</template>
`;

exports[`htmlparser2 option parses XHTML with xmlMode enabled 1`] = `
<ul>
  <li />
  <li />
</ul>
`;

exports[`replace option does not replace the element if an invalid React element is returned 1`] = `
<html>
  <head>
    <meta
      charSet="utf-8"
    />
    <title>
      Title
    </title>
    <link
      href="style.css"
      rel="stylesheet"
    />
  </head>
  <body>
    <header
      id="header"
    >
      Header
    </header>
    <h1
      style={
        {
          "color": "#000",
          "fontSize": "42px",
        }
      }
    >
      Heading
    </h1>
    <hr />
    <p>
      Paragraph
    </p>
    <img
      src="image.jpg"
    />
    <div
      className="class1 class2"
    >
      Some 
      <em>
        text
      </em>
      .
    </div>
    <script
      dangerouslySetInnerHTML={
        {
          "__html": "alert();",
        }
      }
    />
  </body>
</html>
`;

exports[`replace option replaces the element if a valid React element is returned 1`] = `
<html>
  <head>
    <meta
      charSet="utf-8"
    />
    <title>
      Replaced Title
    </title>
    <link
      href="style.css"
      rel="stylesheet"
    />
  </head>
  <body>
    <header
      id="header"
    >
      Header
    </header>
    <h1
      style={
        {
          "color": "#000",
          "fontSize": "42px",
        }
      }
    >
      Heading
    </h1>
    <hr />
    <p>
      Paragraph
    </p>
    <img
      src="image.jpg"
    />
    <div
      className="class1 class2"
    >
      Some 
      <em>
        text
      </em>
      .
    </div>
    <script
      dangerouslySetInnerHTML={
        {
          "__html": "alert();",
        }
      }
    />
  </body>
</html>
`;

exports[`trim option preserves whitespace text nodes when disabled if valid in parent (default) 1`] = `
<table>
  <tbody>
    <tr>
      <td>
        hello
      </td>
      <td>
        

      </td>
      <td>
        Â 
      </td>
    </tr>
  </tbody>
</table>
`;
